xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,251 @@
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.XdsClient.ClusterUpdate;
+import io.grpc.xds.XdsClient.ClusterWatcher;
+import io.grpc.xds.XdsLoadBalancerProvider.XdsConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_cds LB policy.
+ */
+public final class CdsLoadBalancer extends LoadBalancer {
+  private final ChannelLogger channelLogger;
+  private final LoadBalancerRegistry lbRegistry;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
+
+  // The following fields become non-null once handleResolvedAddresses() successfully.
+  /** Most recent XdsConfig. */
+  @Nullable
+  private CdsConfig cdsConfig;
+  /** Most recent ClusterWatcher. */
+  @Nullable
+  private ClusterWatcher clusterWatcher;
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  @Nullable
+  private XdsClient xdsClient;
+
+  CdsLoadBalancer(Helper helper) {
+    this(helper, LoadBalancerRegistry.getDefaultRegistry());
+  }
+
+  @VisibleForTesting
+  CdsLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry) {
+    this.channelLogger = helper.getChannelLogger();
+    this.lbRegistry = lbRegistry;
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(helper);
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+    Object lbConfig = resolvedAddresses.getLoadBalancingPolicyConfig();
+    Preconditions.checkArgument(
+        lbConfig != null && CdsConfig.class.isAssignableFrom(lbConfig.getClass()),
+        "Expecting a CDS LB config, but the actual LB config is '%s'",
+        lbConfig);
+
+    if (xdsClientRef == null) {
+      xdsClientRef = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_REF);
+
+      if (xdsClientRef == null) {
+        // TODO(zdapeng): create a new xdsClient from bootstrap if no one exists.
+        throw new UnsupportedOperationException(
+            "XDS_CLIENT_REF attributes not available in resolvedAddresses " + resolvedAddresses);
+      }
+
+      xdsClient = xdsClientRef.getObject();
+    }
+
+    final CdsConfig newCdsConfig = (CdsConfig) lbConfig;
+    // If CdsConfig is changed, do a graceful switch.
+    if (!newCdsConfig.equals(cdsConfig)) {
+      final CdsConfig oldCdsConfig = cdsConfig;
+      final ClusterWatcher oldClusterWatcher = clusterWatcher;
+      // Provides a load balancer with the fixed CdsConfig: newCdsConfig.
+      LoadBalancerProvider fixedCdsConfigBalancer = new LoadBalancerProvider() {
+        ClusterWatcherImpl clusterWatcher;
```

Comment:
Wouldn't it be okay to initialize this with `CdsLoadBalancer.this.clusterWatcher` ?